query-engine/core/src/query_graph_builder/write/utils.rs

```diff
@@ -515,16 +518,17 @@ pub fn emulate_on_delete_cascade(
     graph: &mut QueryGraph,
     relation_field: &RelationFieldRef, // This is the field _on the other model_ for cascade.
     query_schema: &QuerySchema,
-    parent_node: &NodeRef,
+    parent_node: Option<&NodeRef>,
     child_node: &NodeRef,
 ) -> QueryGraphBuilderResult<()> {
     let dependent_model = relation_field.model();
     let parent_relation_field = relation_field.related_field();
     let child_model_identifier = parent_relation_field.related_model().primary_identifier();
 
     // Records that need to be deleted for the cascade.
+    let node_providing_ids = parent_node.unwrap_or(child_node);
```

Comment:
Is the idea here that if the parent is None, it means the delete is atomic and therefore we can use the child node (which is the atomic delete node) as a source node to attach the children read? 🤔

Reply:
Exactly. This made function's semantics too complicated, I think. I simplified it's interface in the latest patch to only create the emulation subtree. Creation of execution order edges now lives in its own separate function. This way, the notion of parent/children nodes goes away in this context, we only have the node which provides the ids we need to delete.